flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinator.java

```diff
@@ -201,8 +201,11 @@ void announceCombinedWatermark() {
         // to ready task to avoid period task fail (Java-ThreadPoolExecutor will not schedule
         // the period task if it throws an exception).
         for (Integer subtaskId : subTaskIds) {
-            context.sendEventToSourceOperatorIfTaskReady(
-                    subtaskId, new WatermarkAlignmentEvent(maxAllowedWatermark));
+            // when subtask have been finished, do not send event.
+            if (!context.hasNoMoreSplits(subtaskId)) {
```

Comment:
> I believe these two locations should be equivalent.

I think you are right, both of `DataInputStatus.END_OF_INPUT` and `NoMoreSplitsEvent` work well for now.

I prefer `DataInputStatus.END_OF_INPUT` because its semantics are clearer, and it's closer to FINISHED. It means the task will be definitely switched to FINISHED after `DataInputStatus.END_OF_INPUT`.

As we all know, code is often refactored. We shouldn't make assumptions (especially assumptions that other developers don't know about).  Other developers may break our assumption if they don't know during the refactor in the future. I'm afraid the `NoMoreSplitsEvent` is used for other cases, and task doesn't switch to FINISHED after receiving `NoMoreSplitsEvent` in the future. That's my concern.

Reply:
BTW, would you mind adding the demo from FLINK-35157 as an ITCase to check whether the watermark alignment works well with finished task?

This ITCase will fail or timeout(as described by FLINK-35157: source is blocked) if another developer breaks this feature accidentally. WDYT?